### PR TITLE
[chore] Make comparetest and pdatautil packages exported under pkg

### PR DESCRIPTION
**Pull request by dmitryax**
Move `internal/coreinternal/pdatautil` and `internal/comparetest` module in pkg and export them as `pkg/pdatautil` and `pkg/pdatatest` so they can be used outside of the contrib repository after `pcommon.Map.Sort()` method is deprecated.

Once stabilized, they can be moved to the core pdata module.

Updates https://github.com/open-telemetry/opentelemetry-collector/issues/6688

### CODE REVIEW

**Review by mx-psi — COMMENTED** [review]
What modules would use this outside of contrib?

**Comment by dmitryax** [comment]
> What modules would use this outside of contrib?

Any other distros that use Sort() to do things like Logs == Logs in tests or making signatures from the attributes map

**Comment by dmitryax** [comment]
Two examples where the pdatatest is needed after Map.Sort() is removed:
- https://github.com/influxdata/influxdb-observability/blob/4e30db9141147d5b37ec2967baaf17053cd86baf/influx2otel/metrics_helper_test.go#L13
- https://github.com/SumoLogic/sumologic-otel-collector/blob/547af6af647bf09b0dd652fa2f6a95ae4be56fc1/pkg/processor/sourceprocessor/source_processor_test.go#L136

**Review by bogdandrutu — COMMENTED** [review]
1. Should the pdatatest follow the same pdata structure and have a `ptracetest`, etc.?
2. Where is the golden used? Do we need to expose that now?
3. Should pdatautil actually be `pcommonutil`?

**Comment by dmitryax** [comment]
> Where is the golden used? Do we need to expose that now?

Probably not needed to be exposed now. I'll move it aside.

> Should the pdatatest follow the same pdata structure and have a ptracetest, etc.?

That makes sense to keep the package name the same as it'll eventually land in core/pdata, just to reduce the amount of work for users when we will be moving it to core, so they need to rename the module path, but not the package name. I'll update it.

> Should pdatautil actually be pcommonutil?

I don't think we will need `pcommonutil` in core for the current functionality. I think it'll be `pcommon.[Map|Value].Hash()` methods, so it doesn't matter what's the module name here. I think `pdatautil` is a good one to start with

**Comment by dmitryax** [comment]
@bogdandrutu I've applied the changes discussed above

File: pkg/pdatatest/go.mod [hunk 1]
```diff
@@ -1,9 +1,10 @@
-module github.com/open-telemetry/opentelemetry-collector-contrib/internal/comparetest
+module github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest
```
Comment (bogdandrutu) [inline pkg/pdatatest/go.mod:1]: not sure if the module should be pdatatest or just pdata.

File: pkg/pdatatest/plogtest/logs.go [hunk 2]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package comparetest // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/comparetest"
+package plogtest // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest/plogtest"
```
Comment (bogdandrutu) [inline pkg/pdatatest/plogtest/logs.go:15]: `CompareLogs` is redundant? Just Compare is good enough, or you think of adding things like CompareLogRecord?

File: pkg/pdatatest/plogtest/options.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest/internal"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+// LogsCompareOption can be used to mutate expected and/or actual logs before comparing.
+type LogsCompareOption interface {
```
Comment (bogdandrutu) [inline pkg/pdatatest/plogtest/options.go:29]: Same, the Logs prefix seem a bit redundant.

File: pkg/pdatatest/plogtest/options.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+func IgnoreObservedTimestamp() LogsCompareOption {
+	return ignoreObservedTimestamp{}
+}
+
+type ignoreObservedTimestamp struct{}
+
+func (opt ignoreObservedTimestamp) applyOnLogs(expected, actual plog.Logs) {
+	now := pcommon.NewTimestampFromTime(time.Now())
+	maskObservedTimestamp(expected, now)
+	maskObservedTimestamp(actual, now)
```
Comment (bogdandrutu) [inline pkg/pdatatest/plogtest/options.go:66]: You don't have to have a struct for every options, do something like:

```suggestion
func IgnoreObservedTimestamp() LogsCompareOption {
	return logsCompareOptionFunc(func(expected, actual plog.Logs) {
		now := pcommon.NewTimestampFromTime(time.Now())
		maskObservedTimestamp(expected, now)
		maskObservedTimestamp(actual, now)
	})
}

type logsCompareOptionFunc func(expected, actual plog.Logs)

func (f ignoreObservedTimestamp) applyOnLogs(expected, actual plog.Logs) {
	f(expected, actual)
}
```

File: pkg/pdatatest/go.mod (hunk 1 above)
Comment (dmitryax) [reply pkg/pdatatest/go.mod:1]: I thought pdatatest is better here to not confuse with actual core/pdata module

File: pkg/pdatatest/plogtest/logs.go (hunk 2 above)
Comment (dmitryax) [reply pkg/pdatatest/plogtest/logs.go:15]: We already have Compare to other objects here like `CompareScopeLogs` and  `CompareLogRecords`

File: pkg/pdatatest/plogtest/options.go (hunk 3 above)
Comment (dmitryax) [reply pkg/pdatatest/plogtest/options.go:29]: Currently, this option is applicable to `CompareLogs` only (supposed to be called `CompareLogsOption` btw), `CompareScopeLogs` and `CompareLogRecords` go with no options at this point. If we want to add options to those functions later, we will need `CompareScopeLogsOption` and `CompareLogRecordsOption`. That's why I think it's better to be explicit with the naming like `<function_name>Option`. Let me know WDYT

File: pkg/pdatatest/plogtest/options.go (hunk 4 above)
Comment (dmitryax) [reply pkg/pdatatest/plogtest/options.go:66]: Yes, that was on my list of improvements just didn't have a chance to apply it. I can do it in this PR

Comment (dmitryax) [reply pkg/pdatatest/plogtest/options.go]: Updated

File: pkg/pdatatest/plogtest/options.go (hunk 3 above)
Comment (bogdandrutu) [reply pkg/pdatatest/plogtest/options.go]: > should to be called CompareLogsOption btw

100% agree, prefix with the func name.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**